### PR TITLE
mpl2: support snapping of macros with pins in multiple layers

### LINKED ISSUE #5854: invalid macro placement leads to detailed routing problems with pins on more than one metal layer
### Describe the bug

Standalone bugreports broken out from https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/2412

untar [detail-route-problem-with-pins-on-multiple-metals.tar.gz](https://github.com/user-attachments/files/17232782/detail-route-problem-with-pins-on-multiple-metals.tar.gz)

Detailed routing eventually completes with lots of DRC errors:

```
$ ./run-me-mock-array-asap7-base.sh 
OpenROAD v2.0-16219-g56c3c86ba 
[deleted]
[WARNING DRT-0418] Term ces_0_0/clock has no pins on routing grid
[WARNING DRT-0418] Term ces_0_0/io_ins_down[0] has no pins on routing grid
[WARNING DRT-0418] Term ces_0_0/io_ins_down[10] has no pins on routing grid
[lots]
[deleted]
```

The root cause might be macro placement, broken out as a standalone test here:

[macro-place-multi-metal-pins.tar.gz](https://github.com/user-attachments/files/17232833/macro-place-multi-metal-pins.tar.gz)

```
$ ./run-me-mock-array-asap7-base.sh 
OpenROAD v2.0-16219-g56c3c86ba 
Features included (+) or not (-): +Charts +GPU +GUI +Python
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
HierRTLMP Flow enabled...
rtl_macro_placer -halo_width 0 -halo_height 2.16 -boundary_weight 0 -report_directory ./objects/asap7/mock-array/base/rtlmp -target_util 0.30
Floorplan Outline: (0.0, 0.0) (358.56, 388.8),  Core Outline: (2.16, 2.16) (356.4, 386.64)
	Number of std cell instances: 129
	Area of std cell instances: 23.37
	Number of macros: 64
	Area of macros: 119439.39
	Halo width: 0.00
	Halo height: 2.16
	Area of macros with halos: 131383.38
	Area of std cell instances + Area of macros: 119462.77
	Core area: 136198.20
	Design Utilization: 0.88
	Core Utilization: 0.00
	Manufacturing Grid: 1

openroad> 
```

### Expected Behavior

No error

### Environment

```shell
OpenROAD v2.0-16219-g56c3c86ba
```

### To Reproduce

See above

### Relevant log output

_No response_

### Screenshots

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by AcKoucher**
Resolve #5854.

Attempts to move the macro to different track-grid positions after snapping based on the first layer we find.
First all vertical layers with pins, then all horizontal layers with pins.
The goal is to search for a location that results in the largest number of layers with pins in the same direction to have those pins correctly aligned with the grid.

### CODE REVIEW

File: src/mpl2/src/hier_rtlmp.cpp [hunk 1]
```diff
@@ -4463,186 +4463,252 @@ bool Pusher::overlapsWithIOBlockage(const odb::Rect& cluster_box,
...
+    : logger_(logger), inst_(inst)
 {
 }
 
 void Snapper::snapMacro()
 {
-  const odb::Point snap_origin = computeSnapOrigin();
-  inst_->setOrigin(snap_origin.x(), snap_origin.y());
+  snap(true /*horizontal snap*/);
+  snap(false /*horizontal snap*/);
```
Comment (eder-matheus) [inline src/mpl2/src/hier_rtlmp.cpp:4478]: Shouldn't it be `vertical snap`?

File: src/mpl2/src/hier_rtlmp.cpp [hunk 2]
```diff
@@ -4463,186 +4463,252 @@ bool Pusher::overlapsWithIOBlockage(const odb::Rect& cluster_box,
...
+        all_layers_snapped = true;
+        break;
+      }
+    }
+
+    if (all_layers_snapped) {
+      return;
+    }
+
+    if (curr_number_of_snapped_layers > best_number_of_snapped_layers) {
```
Comment (eder-matheus) [inline src/mpl2/src/hier_rtlmp.cpp:4695]: You should update the `best_number_of_snapped_layers` with the new number of snapped layers.

File: src/mpl2/src/hier_rtlmp.cpp [hunk 3]
```diff
@@ -4463,186 +4463,252 @@ bool Pusher::overlapsWithIOBlockage(const odb::Rect& cluster_box,
...
+  int remaining_attempts = total_attempts;
+
+  int best_origin = origin;
+  int best_number_of_snapped_layers = 1;
+  int new_origin = origin;
+  bool all_layers_snapped = false;
+
+  while (remaining_attempts > 0) {
+    // Move one track ahead.
+    new_origin += snap_layer_params.pitch;
```
Comment (eder-matheus) [inline src/mpl2/src/hier_rtlmp.cpp:4671]: Here you only move the origin considering the pitch of the same layer in every iteration?

**Review by eder-matheus — CHANGES_REQUESTED** [review]

File: src/mpl2/src/hier_rtlmp.cpp (hunk 1 above)
Comment (AcKoucher) [reply src/mpl2/src/hier_rtlmp.cpp:4478]: Afaiu, we snap horizontally to align the pins of a layer whose preferred direction is vertical:

https://github.com/The-OpenROAD-Project/OpenROAD/blob/934fad8a8cf47f56cccd918da3ef0160447a4433/src/mpl2/src/hier_rtlmp.cpp#L4533-L4539

File: src/mpl2/src/hier_rtlmp.cpp (hunk 3 above)
Comment (AcKoucher) [reply src/mpl2/src/hier_rtlmp.cpp:4671]: Yes, the snap_layer is the layer that we "manually" snapped. So we search for alignment in other layers by jumping its tracks.

File: src/mpl2/src/hier_rtlmp.cpp (hunk 1 above)
Comment (AcKoucher) [reply src/mpl2/src/hier_rtlmp.cpp:4478]: Oh, I see. My intentions were to replicate the boolean name, but apparently it didn't make it very clear. I'll write exaclty what each one is doing:
```
  snap(true); // Horizontal snap
  snap(false); // Vertical snap

File: src/mpl2/src/hier_rtlmp.cpp [hunk 4]
```diff
@@ -4463,186 +4463,253 @@ bool Pusher::overlapsWithIOBlockage(const odb::Rect& cluster_box,
...
+    : logger_(logger), inst_(inst)
 {
 }
 
 void Snapper::snapMacro()
 {
-  const odb::Point snap_origin = computeSnapOrigin();
-  inst_->setOrigin(snap_origin.x(), snap_origin.y());
+  snap(true);   // Horizontal snap
+  snap(false);  // Vertical snap
```
[the remark this answers is not in the document]
Comment (AcKoucher) [reply src/mpl2/src/hier_rtlmp.cpp:4478]: Yes. That will make it clearer.

**Comment by oharboe** [comment]
Took it for a quick spin on https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/2412:

```
make DESIGN_CONFIG=designs/asap7/mock-array/config.mk
```

Results in error:

```
[INFO PPL-0007] Random pin placement.
[WARNING PPL-0106] At least 2 pins in position (42633, 43200), layer M3, port io_ins_down[61].
[ERROR PPL-0107] Invalid pin placement.
Error: io_placement_util.tcl, 11 PPL-0107
```

**Comment by oharboe** [comment]
> @oharboe I believe this is not related to the macro placement. Do you see macros with pins out of track after mpl?

I don't know how to check that...

**Comment by AcKoucher** [comment]
You can set the tracks visibility in the layout display and then find the pins in the layers you want.

Without the changes here, the pins in some layers were not properly aligned with the grid.
With the changes here we have something like this:

image

So you shouldn't have the DRCs problems anymore.

**Comment by oharboe** [comment]
Never gets far enough... This might be another manifestation of a problem that I raised with @maliberty ?

Standalone testcase [io_placement_problem.tar.gz](https://github.com/user-attachments/files/17284662/io_placement_problem.tar.gz)


```
$ make DESIGN_CONFIG=designs/asap7/mock-array/config.mk 
Running io_placement_random.tcl, stage 2_2_floorplan_io
[INFO ORD-0030] Using 16 thread(s).
[INFO PPL-0048] Restrict pins [ io_ins_up[0] io_ins_up[10] io_ins_up[11] io_ins_up[12] io_ins_up[13] ... ] to region 0.00u-43.20u at the BOTTOM edge.
[INFO PPL-0048] Restrict pins [ io_ins_left[0] io_ins_left[10] io_ins_left[11] io_ins_left[12] io_ins_left[13] ... ] to region 0.00u-43.20u at the RIGHT edge.
[INFO PPL-0048] Restrict pins [ io_lsbOuts_0 io_lsbOuts_1 io_lsbOuts_2 io_lsbOuts_3 io_lsbOuts_4 ... ] to region 0.00u-43.20u at the RIGHT edge.
[INFO PPL-0048] Restrict pins [ clock ] to region 0.00u-43.20u at the TOP edge.
place_pins -hor_layers M2 M4 -ver_layers M3 M5 -random
Found 0 macro blocks.
Using 2 tracks default min distance between IO pins.
[INFO PPL-0007] Random pin placement.
[WARNING PPL-0106] At least 2 pins in position (42633, 43200), layer M3, port io_ins_down[61].
[ERROR PPL-0107] Invalid pin placement.
Error: io_placement_util.tcl, 11 PPL-0107
Command exited with non-zero status 1
Elapsed time: 0:01.09[h:]min:sec. CPU time: user 1.04 sys 0.05 (100%). Peak memory: 217100KB.
make[2]: *** [Makefile:651: do-2_2_floorplan_io] Error 1
make[1]: *** [Makefile:649: results/asap7/mock-array_Element/base/2_2_floorplan_io.odb] Error 2
make: *** [Makefile:351: results/asap7/mock-array_Element/base/Element.lib] Error 2
$ ls results/asap7/mock-array_Element/base/
1_1_yosys.v  1_synth.rtlil  1_synth.sdc  1_synth.v  2_1_floorplan.odb  2_floorplan.sdc  clock_period.txt  keep_hierarchy.tcl  mem.json
```

**Comment by eder-matheus** [comment]
@oharboe @AcKoucher The PPL error was fixed last week, but your branch is not updated with the latest master branch. You should not see the PPL issue once you merge the latest master in this PR.

**Comment by oharboe** [comment]
> @oharboe @AcKoucher The PPL error was fixed last week, but your branch is not updated with the latest master branch. You should not see the PPL issue once you merge the latest master in this PR.

Indeed.

**Comment by oharboe** [comment]
https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/2412 seems to work fine with this PR+master of OpenROAD. :+1:

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
secure CI?

**Comment by maliberty** [comment]
@eder-matheus you still have requested change status

**Review by eder-matheus — APPROVED** [review]

**Comment by AcKoucher** [comment]
Running secure-CI

**Comment by AcKoucher** [comment]
@eder-matheus @maliberty 

Public CI showed some failures:

```
nangate45 bp_be
      [ERROR DRT-0416] Term w_mask_in[0] contains offgrid pin shape. Pin shape ( 1070263 1136310 ) ( 1070403 1136450 ) is not a multiple of the manufacturing grid 10.

nangate45 bp_fe
      [ERROR GRT-0119] Routing congestion too high. Check the congestion heatmap in the GUI and load ./reports/nangate45/bp_fe/base/congestion.rpt in the DRC viewer.

nangate45 bp_multi
      [ERROR DRT-0416] Term w_mask_in[0] contains offgrid pin shape. Pin shape ( 1872632 45150 ) ( 1872772 45290 ) is not a multiple of the manufacturing grid 10.

nangate45 tinyRocket
      [ERROR DRT-0416] Term w_mask_in[0] contains offgrid pin shape. Pin shape ( 142613 829430 ) ( 142753 829570 ) is not a multiple of the manufacturing grid 10.
```
These were because for macros that don't have pins to be aligned with the track in a certain direction, we were not aligning with the manufacturing grid.

I pushed a change to ensure that, even we don't need to align to the tracks, we'll still align with the manufacturing grid.

I'll re-run CI.

**Comment by AcKoucher** [comment]
Secure-CI is fine.

**Pull request merged**